
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="Documentation for running Bioinformatics/Computational workflows with bioflowa">
      
      
      
        <meta name="author" content="Ashok Ragavendran">
      
      
        <link rel="canonical" href="https://compbiocore.github.io/yaml_description/">
      
      <link rel="shortcut icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.1.2, mkdocs-material-6.2.7">
    
    
      
        <title>Specifications - Computational Biology Core - Brown University</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.cb6bc1d0.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.39b8e14a.min.css">
        
          
          
          <meta name="theme-color" content="#546d78">
        
      
    
    
    
      
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700%7CRoboto+Mono&display=fallback">
        <style>body,input{font-family:"Roboto",-apple-system,BlinkMacSystemFont,Helvetica,Arial,sans-serif}code,kbd,pre{font-family:"Roboto Mono",SFMono-Regular,Consolas,Menlo,monospace}</style>
      
    
    
    
      <link rel="stylesheet" href="../styles/dark_mode.css">
    
      <link rel="stylesheet" href="../docs/assets/adm.css">
    
      <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.8.0/css/all.css">
    
    
      
        
<script>window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)},ga.l=+new Date,ga("create","UA-115983496-2","compbiocore.github.io"),ga("set","anonymizeIp",!0),ga("send","pageview"),document.addEventListener("DOMContentLoaded",function(){document.forms.search&&document.forms.search.query.addEventListener("blur",function(){if(this.value){var e=document.location.pathname;ga("send","pageview",e+"?q="+this.value)}})}),document.addEventListener("DOMContentSwitch",function(){ga("send","pageview",document.location.pathname)})</script>
<script async src="https://www.google-analytics.com/analytics.js"></script>
      
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="" data-md-color-primary="blue-grey" data-md-color-accent="indigo">
      
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#bioflows-yaml-control-file-specifications" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header-nav md-grid" aria-label="Header">
    <a href="https://compbiocore.github.io/" title="Computational Biology Core - Brown University" class="md-header-nav__button md-logo" aria-label="Computational Biology Core - Brown University">
      
  <img src="../assets/cbc-logo.svg" alt="logo">

    </a>
    <label class="md-header-nav__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2z"/></svg>
    </label>
    <div class="md-header-nav__title" data-md-component="header-title">
      <div class="md-header-nav__ellipsis">
        <div class="md-header-nav__topic">
          <span class="md-ellipsis">
            Computational Biology Core - Brown University
          </span>
        </div>
        <div class="md-header-nav__topic">
          <span class="md-ellipsis">
            
              Specifications
            
          </span>
        </div>
      </div>
    </div>
    
      <label class="md-header-nav__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0116 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 019.5 16 6.5 6.5 0 013 9.5 6.5 6.5 0 019.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
      </label>
      
<div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" data-md-state="active" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0116 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 019.5 16 6.5 6.5 0 013 9.5 6.5 6.5 0 019.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
      </label>
      <button type="reset" class="md-search__icon md-icon" aria-label="Clear" data-md-component="search-reset" tabindex="-1">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"/></svg>
      </button>
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header-nav__source">
        
<a href="https://github.com/compbiocore/bioflows/" title="Go to repository" class="md-source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05L244 40.45a28.87 28.87 0 00-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 01-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 000 40.81l195.61 195.6a28.86 28.86 0 0040.8 0l194.69-194.69a28.86 28.86 0 000-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    




<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="https://compbiocore.github.io/" title="Computational Biology Core - Brown University" class="md-nav__button md-logo" aria-label="Computational Biology Core - Brown University">
      
  <img src="../assets/cbc-logo.svg" alt="logo">

    </a>
    Computational Biology Core - Brown University
  </label>
  
    <div class="md-nav__source">
      
<a href="https://github.com/compbiocore/bioflows/" title="Go to repository" class="md-source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05L244 40.45a28.87 28.87 0 00-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 01-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 000 40.81l195.61 195.6a28.86 28.86 0 0040.8 0l194.69-194.69a28.86 28.86 0 000-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        Home
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../intro/" class="md-nav__link">
        Docs
      </a>
    </li>
  

    
      
      
      

  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          Specifications
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        Specifications
      </a>
      
        
<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#project-information-section" class="md-nav__link">
    Project information section
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#global-run-parameters" class="md-nav__link">
    Global run parameters
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#workflow-parameters" class="md-nav__link">
    Workflow parameters
  </a>
  
    <nav class="md-nav" aria-label="Workflow parameters">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#how-to-run" class="md-nav__link">
    How to run
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="nav-4" type="checkbox" id="nav-4" >
      
      <label class="md-nav__link" for="nav-4">
        Tutorials
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Tutorials" data-md-level="1">
        <label class="md-nav__title" for="nav-4">
          <span class="md-nav__icon md-icon"></span>
          Tutorials
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../tutorials/tutorials/" class="md-nav__link">
        About
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../tutorials/Setup_bioflows_env/" class="md-nav__link">
        Setup the bioflows environment
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../tutorials/rna-seq_tutorial/" class="md-nav__link">
        RNA-Seq tutorial
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../tutorials/qiime2_tutorial/" class="md-nav__link">
        Qiime2 tutorial
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    
<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#project-information-section" class="md-nav__link">
    Project information section
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#global-run-parameters" class="md-nav__link">
    Global run parameters
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#workflow-parameters" class="md-nav__link">
    Workflow parameters
  </a>
  
    <nav class="md-nav" aria-label="Workflow parameters">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#how-to-run" class="md-nav__link">
    How to run
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content">
            <article class="md-content__inner md-typeset">
              
                
                  <a href="https://github.com/compbiocore/bioflows/edit/master/docs/yaml_description.md" title="Edit this page" class="md-content__button md-icon">
                    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25z"/></svg>
                  </a>
                
                
                <h1 id="bioflows-yaml-control-file-specifications"><code>bioflows</code> YAML control file Specifications</h1>
<h2 id="project-information-section">Project information section</h2>
<ul>
<li>
<p><code>bioproject</code>: This will be an unique identifier for your project.
    This is adopted from the NCBI SRA format structure, so if you use an
    SRA dataset you can employ these ids</p>
</li>
<li>
<p><code>experiment</code>: An identifier for your experiment such as RNA-seq,
    ChIP-seq etc</p>
</li>
<li>
<p><code>sample_manifest</code>: This section contains two sections</p>
<ul>
<li>
<p><code>fastq_file</code>: The full path to the sample to fastq map file.
    This file is in a three column comma separated format with each line formatted as:</p>
<p><code>sample_id, path_to_fastq_file_for_read1, path_to_fastq_file_for_read2</code></p>
<p>The <code>sample_id</code> is unique and if you are <strong>using single end data</strong> you just need
to specify one column as shown below:</p>
<p><code>sample_id, path_to_fastq_file</code></p>
</li>
<li>
<p><code>metadata</code>: This is all the metadata associated with a given
    <code>sample_id</code> if available such as gender, extraction date etc. This
    should also be a CSV format file. Currently, not necessary as this
    information is not yet used</p>
</li>
</ul>
</li>
</ul>
<h2 id="global-run-parameters">Global run parameters</h2>
<ul>
<li>
<p><code>run_parms</code>: This section specifies the global parameters for the
    current analysis</p>
<ul>
<li>
<p><code>conda_command</code>: This is the command used to activate your conda
    environment</p>
</li>
<li>
<p><code>work_dir</code>: The working directory for analysis usually created on  <code>/gpfs/scratch</code></p>
</li>
<li>
<p><code>log_dir</code>: The subdirectory for all the log files</p>
</li>
<li>
<p><code>paired_end</code>: Whether data consists of paired end reads or single end
    reads (True/False)</p>
</li>
</ul>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This is an experimental feature and may not work as intended</p>
</div>
<ul>
<li>
<p><code>ssh_user</code>: The user name if workflow is run from a local machine</p>
</li>
<li>
<p><code>saga_scheduler</code>: The scheduler being used, for CCV the value
    used here is slurm. 
    !!! note 
        Currently only tested with <code>slurm</code> scheduler. Will add test to others soon</p>
</li>
<li>
<p><code>gtf_file</code>: The full path to the gtf file for gene annotations, needed if you are planning to run rna-seq analysis</p>
</li>
</ul>
</li>
</ul>
<h2 id="workflow-parameters">Workflow parameters</h2>
<ul>
<li>
<p><code>workflow_sequence</code>: This section specifies the sequence of tools to
    be used and the options passed to tools as well as the job parameters
    if using a scheduler such as slurm</p>
<ul>
<li>
<p><code>fastqc</code>:If you want to use the default parameters use <em>default</em>
    else you can use any of the options provided by the
    program. See  the example for GSNAP below on how to do
    that. See the documentation for the options for fastqc.</p>
</li>
<li>
<p><code>gsnap</code>: Here we give an example of two sections as we need to
    pass the index information to the aligner</p>
<ul>
<li>
<p><code>options</code>: Specify program options here. In this example we
    specify the following</p>
<ul>
<li>
<p><strong>-d:</strong> The genome index for GSNAP</p>
</li>
<li>
<p><strong>-s:</strong> and the splicesites file location for GSNAP.</p>
</li>
</ul>
<p>The format is exactly that as to what you would specify on the command line for the program</p>
<div class="codehilite"><pre><span></span><code><span class="err">-d Ensembl_mus_GRCm38.p5_rel89</span>
<span class="err">-s Mus_musculus.GRCm38.89.splicesites.iit</span>
</code></pre></div>

<p>See the documentation for the GSNAP program for more options</p>
</li>
<li>
<p><code>job_param</code>s: This section specifies parameters for job submission such as memory, number of cores etc</p>
<ul>
<li>ncpus: 16</li>
<li>mem: 40000</li>
<li>time: 60</li>
</ul>
</li>
</ul>
</li>
<li>
<p><code>qualimap_rnaseq</code>: Run the qualimap module for RNAseq with the <strong>default</strong> settings</p>
</li>
</ul>
</li>
</ul>
<p>The final YAML control file should look as below to run a test example. Only modify the parts
that are highlighted below to fill in your own values.</p>
<div class="codehilite"><pre><span></span><code><span class="n">bioproject</span><span class="o">:</span> <span class="n">Project_test_localhost</span>
<span class="n">experiment</span><span class="o">:</span> <span class="n">rnaseq_pilot</span>
<span class="n">sample_manifest</span><span class="o">:</span>
  <span class="n">fastq_file</span><span class="o">:</span> <span class="sr">/users/:bluetext:`username/s</span><span class="n">ample_manifest_min</span><span class="o">.</span><span class="na">csv</span><span class="err">`</span>
  <span class="n">metadata</span><span class="o">:</span>
<span class="n">run_parms</span><span class="o">:</span>
  <span class="n">conda_command</span><span class="o">:</span> <span class="n">source</span> <span class="n">activate</span> <span class="sr">/gpfs/runtime/opt/conda/envs/</span><span class="n">cbc_conda_test</span>
  <span class="n">work_dir</span><span class="o">:</span> <span class="sr">/gpfs/scratch/&#39;user/</span><span class="n">test_workflow</span><span class="s1">&#39;</span>
<span class="s1">  log_dir: logs</span>
<span class="s1">  paired_end: False</span>
<span class="s1">  local_targets: False</span>
<span class="s1">  db: sqlite</span>
<span class="s1">  db_loc: &quot;:memory:&quot;</span>
<span class="s1">  saga_host: localhost</span>
<span class="s1">  ssh_user: &#39;</span><span class="n">ccv</span> <span class="n">username</span><span class="err">&#39;</span>
  <span class="n">saga_scheduler</span><span class="o">:</span> <span class="n">slurm</span>
  <span class="n">gtf_file</span><span class="o">:</span> <span class="sr">/gpfs/data/cbc/cbcollab/ref_tools/Ensembl_hg_GRCh37_rel87/</span><span class="n">Homo_sapiens</span><span class="o">.</span><span class="na">GRCh37</span><span class="o">.</span><span class="mi">87</span><span class="o">.</span><span class="na">gtf</span>
<span class="n">workflow_sequence</span><span class="o">:</span>
  <span class="n">fastqc</span><span class="o">:</span> <span class="k">default</span>
  <span class="n">gsnap</span><span class="o">:</span>
    <span class="n">options</span><span class="o">:</span>
      <span class="o">-</span><span class="n">d</span><span class="o">:</span> <span class="n">Ensembl_Homo_sapiens_GRCh37</span>
      <span class="o">-</span><span class="n">s</span><span class="o">:</span> <span class="n">Ensembl_Homo_sapiens</span><span class="o">.</span><span class="na">GRCh37</span><span class="o">.</span><span class="mi">87</span><span class="o">.</span><span class="na">splicesites</span><span class="o">.</span><span class="na">iit</span>
    <span class="n">job_params</span><span class="o">:</span>
      <span class="n">ncpus</span><span class="o">:</span> <span class="mi">16</span>
      <span class="n">mem</span><span class="o">:</span> <span class="mi">40000</span>
    <span class="n">time</span><span class="o">:</span> <span class="mi">60</span>
  <span class="n">qualimap_rnaseq</span><span class="o">:</span> <span class="k">default</span>
  <span class="n">htseq</span><span class="o">-</span><span class="n">count</span><span class="o">:</span> <span class="k">default</span>
</code></pre></div>

<h3 id="how-to-run">How to run</h3>
<p>Copy the above into a text file and save it in <code>/users/username</code> as <code>test_run.yaml</code></p>
<p>Copy the manifest below into a text file and save it in <code>/users/username</code> as <code>sample_manifest_min.csv</code></p>
<div class="highlight"><pre><span></span><code><span class="l l-Scalar l-Scalar-Plain">samp_1299,/gpfs/scratch/aragaven/rnaseq_test/PE_hg/Cb2_1.gz,/gpfs/scratch/aragaven/rnaseq_test/PE_hg/Cb2_2.gz</span>
<span class="l l-Scalar l-Scalar-Plain">samp_1214,/gpfs/scratch/aragaven/rnaseq_test/PE_hg/Cb_1.gz,/gpfs/scratch/aragaven/rnaseq_test/PE_hg/Cb_2.gz</span>
</code></pre></div>
<p>Now in your screen session run the following commands to setup your
environment if you have not done so previously during the setup or you
have started a new screen session
<div class="highlight"><pre><span></span><code>source activate bflows
bioflows-rnaseq test_run.yaml
</code></pre></div></p>
<p>In this case I have created a small test dataset with 10000 reads from a
test human RNAseq data, so it should run within the hour and you should
see that the alignments are completed.</p>
                
              
              
                


              
            </article>
          </div>
        </div>
      </main>
      
        
<footer class="md-footer">
  
    <div class="md-footer-nav">
      <nav class="md-footer-nav__inner md-grid" aria-label="Footer">
        
          <a href="../intro/" class="md-footer-nav__link md-footer-nav__link--prev" rel="prev">
            <div class="md-footer-nav__button md-icon">
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
            </div>
            <div class="md-footer-nav__title">
              <div class="md-ellipsis">
                <span class="md-footer-nav__direction">
                  Previous
                </span>
                Docs
              </div>
            </div>
          </a>
        
        
          <a href="../tutorials/tutorials/" class="md-footer-nav__link md-footer-nav__link--next" rel="next">
            <div class="md-footer-nav__title">
              <div class="md-ellipsis">
                <span class="md-footer-nav__direction">
                  Next
                </span>
                About
              </div>
            </div>
            <div class="md-footer-nav__button md-icon">
              <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4z"/></svg>
            </div>
          </a>
        
      </nav>
    </div>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
        Made with
        <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
          Material for MkDocs
        </a>
      </div>
      
    </div>
  </div>
</footer>
      
    </div>
    
      <script src="../assets/javascripts/vendor.53cc9318.min.js"></script>
      <script src="../assets/javascripts/bundle.e9c9f54f.min.js"></script><script id="__lang" type="application/json">{"clipboard.copy": "Copy to clipboard", "clipboard.copied": "Copied to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.placeholder": "Type to start searching", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.term.missing": "Missing"}</script>
      
      <script>
        app = initialize({
          base: "..",
          features: [],
          search: Object.assign({
            worker: "../assets/javascripts/worker/search.9c0e82ba.min.js"
          }, typeof search !== "undefined" && search)
        })
      </script>
      
    
  </body>
</html>